pkg/reconciler/ingress/resources/gateway.go

```diff
@@ -445,3 +469,101 @@ func UpdateGateway(gateway *v1beta1.Gateway, want []*istiov1beta1.Server, existi
 func isPlaceHolderServer(server *istiov1beta1.Server) bool {
 	return cmp.Equal(server, &placeholderServer, protocmp.Transform())
 }
+
+// QualifiedGatewayNamesFromContext get gateway names from context.
+func QualifiedGatewayNamesFromContext(ctx context.Context, ing *v1alpha1.Ingress) (map[v1alpha1.IngressVisibility]sets.Set[string], error) {
+	ret := make(map[v1alpha1.IngressVisibility]sets.Set[string])
+
+	gateways, err := gatewaysFromContext(ctx, ing)
+	if err != nil {
+		return ret, fmt.Errorf("failed to get gateways from configuration: %w", err)
+	}
+
+	for _, visibility := range []v1alpha1.IngressVisibility{v1alpha1.IngressVisibilityClusterLocal, v1alpha1.IngressVisibilityExternalIP} {
+		ret[visibility] = sets.New[string]()
+
+		for _, gtw := range gateways[visibility] {
+			ret[visibility] = ret[visibility].Insert(gtw.QualifiedName())
+		}
+	}
+
+	return ret, nil
+}
+
+// gatewaysFromContext get gateways relevant to this ingress from context.
+func gatewaysFromContext(ctx context.Context, ing *v1alpha1.Ingress) (map[v1alpha1.IngressVisibility][]config.Gateway, error) {
+	ret := make(map[v1alpha1.IngressVisibility][]config.Gateway)
+
+	istioConfig := config.FromContext(ctx).Istio
+
+	// External gateways selection
+	externalGateways, err := filterGateway(istioConfig.IngressGateways, ing.ObjectMeta.Labels)
+	if err != nil {
+		return ret, fmt.Errorf("failed to filter external gateways: %w", err)
+	}
+
+	if len(externalGateways) == 0 {
+		externalGateways = istioConfig.DefaultExternalGateways()
+	}
+
+	ret[v1alpha1.IngressVisibilityExternalIP] = externalGateways
+
+	// Local gateways selection
+	localGateways, err := filterGateway(istioConfig.LocalGateways, ing.ObjectMeta.Labels)
+	if err != nil {
+		return ret, fmt.Errorf("failed to filter local gateways: %w", err)
+	}
+
+	if len(localGateways) == 0 {
+		localGateways = istioConfig.DefaultLocalGateways()
+	}
+
+	ret[v1alpha1.IngressVisibilityClusterLocal] = localGateways
+
+	return ret, nil
+}
+
+func filterGateway(gtws []config.Gateway, ingressLabels map[string]string) ([]config.Gateway, error) {
+	ret := make([]config.Gateway, 0, 1)
+
+	for _, gtw := range gtws {
+		if gtw.LabelSelector == nil { // default value
+			continue
+		}
+
+		selector, err := metav1.LabelSelectorAsSelector(gtw.LabelSelector)
+		if err != nil {
+			return ret, fmt.Errorf("failed to create selector from gateway (%s) label selector: %w", gtw.QualifiedName(), err)
+		}
+
+		if !selector.Matches(fields.Set(ingressLabels)) {
+			continue
+		}
+
+		ret = append(ret, gtw)
+	}
+
+	return ret, nil
+}
+
+func PublicGatewayServiceURLFromContext(ctx context.Context, ing *v1alpha1.Ingress) (string, error) {
+	return getGatewayServiceURLFromContext(ctx, ing, v1alpha1.IngressVisibilityExternalIP)
+}
+
+func PrivateGatewayServiceURLFromContext(ctx context.Context, ing *v1alpha1.Ingress) (string, error) {
+	return getGatewayServiceURLFromContext(ctx, ing, v1alpha1.IngressVisibilityClusterLocal)
+}
```

Comment:
These method needs to filter the gateways that apply to this specific ingress. This is why it's not in the config package, it would create a circular dependency

Or, as you mentioned, this method takes as param the list of gateways

I think I will put it back in ingress where it used to be